editor-integration/emacs/dune-mode.el

```diff
@@ -0,0 +1,328 @@
+       "install_c_headers" "modes" "no_dynlink" "kind"
+       "ppx_runtime_libraries" "virtual_deps" "js_of_ocaml" "flags"
+       "ocamlc_flags" "ocamlopt_flags" "library_flags" "c_flags"
+       "cxx_flags" "c_library_flags" "self_build_stubs_archive"
+       "modules_without_implementation"
+       ;; + for "executable" and "executables":
+       "package" "link_flags" "modes" "names" "public_names"
+       ;; + for "rule":
+       "targets" "action" "deps" "mode"
+       ;; + for "menhir":
+       "merge_into"
+       ;; + for "install"
+       "section" "files" "lib" "libexec" "bin" "sbin" "toplevel" "share"
+       "share_root" "etc" "doc" "stublibs" "man" "misc")
+     'symbols))
+  "Field names allowed in dune files.")
+
+(defvar dune-builtin-regex
+  (eval-when-compile
+    (concat (regexp-opt
+             '(;; Actions
+               "run" "chdir" "setenv"
+               "with-stdout-to" "with-stderr-to" "with-outputs-to"
+               "ignore-stdout" "ignore-stderr" "ignore-outputs"
+               "progn" "echo" "write-file" "cat" "copy" "copy#" "system"
+               "bash" "diff" "diff?"
+               ;; inline_tests and inline_tests.backend
+               ;; FIXME: "flags" is already a field and we do not have enough
+               ;; context to distinguishing both.
+               "backend" "generate_runner" "runner_libraries" "flags"
+               "extends")
+             t)
+            "\\(?:\\_>\\|[[:space:]]\\)"))
+  "Builtin sub-fields in dune")
+
+(defvar dune-var-kind-regex
+  (eval-when-compile
+    (regexp-opt
+     '("path" "path-no-dep" "exe" "bin" "lib" "libexec" "lib-available"
+       "version" "read" "read-lines" "read-strings")
+     'words))
+  "Optional prefix to variable names.")
+
+(defvar dune-var-regex
+      (concat "\\(!?\\)\\(\\(?:" dune-var-kind-regex
+              ":\\)?\\)\\([a-zA-Z][a-zA-Z0-9_.-]*\\|[<@^]\\)"
+              "\\(\\(?::[a-zA-Z][a-zA-Z0-9_.-]*\\)?\\)"))
+
+(defmacro dune--field-vals (field &rest vals)
+  `(list (concat "(" ,field "[[:space:]]+" ,(regexp-opt vals t))
+         1 font-lock-constant-face))
+
+(setq dune-font-lock-keywords
+  `((,dune-keywords-regex . font-lock-keyword-face)
+    (,(concat "(" dune-fields-regex) 1 font-lock-function-name-face)
+    ("\\(true\\|false\\)" 1 font-lock-constant-face)
+    ("(\\(select\\)[[:space:]]+[^[:space:]]+[[:space:]]+\\(from\\)\\>"
+     (1 font-lock-constant-face)
+     (2 font-lock-constant-face))
+    ,(eval-when-compile
+       (dune--field-vals "kind" "normal" "ppx_rewriter" "ppx_deriver"))
+    ,(eval-when-compile
+       (dune--field-vals "mode" "standard" "fallback" "promote"
+                                "promote-until-clean"))
+    (,(concat "(" dune-builtin-regex) 1 font-lock-builtin-face)
+    ("(preprocess[[:space:]]+(\\(pps\\)" 1 font-lock-builtin-face)
+    (,(eval-when-compile
+        (concat "(" (regexp-opt '("fallback") t)))
+     1 dune-error-face)
+    (,(concat "${" dune-var-regex "}")
```

Comment:
This is now incorrect — unpublished WIP in my copy.

Reply:
What should I do about this?